### PR TITLE
Uxto tree

### PR DESCRIPTION
**Pull request by oaleksandrov-ip**

### CODE REVIEW

File: sdk/src/main/scala/com/horizen/SidechainState.scala [hunk 1]
```diff
@@ -186,42 +217,48 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage, val
...
+
+    if (closedCoinsBoxesAmount != newCoinsBoxesAmount + tx.fee())
+      throw new Exception("Amounts sum of CoinsBoxes is incorrect. " +
+        s"ClosedBox amount - $closedCoinsBoxesAmount, NewBoxesAmount - $newCoinsBoxesAmount, Fee - ${tx.fee()}")
   }
 
   override def applyModifier(mod: SidechainBlock): Try[SidechainState] = {
     validate(mod).flatMap { _ =>
       changes(mod).flatMap(cs => {
+        closedBoxesMerkleTree.applyBlock(mod)
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/SidechainState.scala:261]: Do we need to update `closedBoxesMerkleTree.applyBlock(mod)` directly here or inside `applyChanges(...)`. How will we react on unexpected failures?

File: sdk/src/main/scala/com/horizen/SidechainState.scala [hunk 2]
```diff
@@ -275,18 +313,34 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage, val
...
 
-  override def rollbackTo(to: VersionTag): Try[SidechainState] = Try {
+  def rollbackTo(to: VersionTag, removedBlocks: Seq[SidechainBlock]): Try[SidechainState] = Try {
     require(to != null, "Version to rollback to must be NOT NULL.")
     val version = BytesUtils.fromHexString(to)
     applicationState.onRollback(version) match {
-      case Success(appState) => new SidechainState(stateStorage.rollback(new ByteArrayWrapper(version)).get, params, to, appState)
+      case Success(appState) => {
+        closedBoxesMerkleTree.removeBlocks(removedBlocks)
+        new SidechainState(stateStorage.rollback(new ByteArrayWrapper(version)).get, params, to, appState, closedBoxesMerkleTree)
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/SidechainState.scala:322]: Maybe we need to do `stateStorage.rollback(new ByteArrayWrapper(version)).get` first, then ` closedBoxesMerkleTree.removeBlocks(removedBlocks)` and then new SidechainState creation.

File: sdk/src/main/scala/com/horizen/forge/ForgeMessageBuilder.scala [hunk 3]
```diff
@@ -221,6 +222,11 @@ class ForgeMessageBuilder(mainchainSynchronizer: MainchainSynchronizer,
           .toSeq
       }
 
+    Try{nodeView.state.findIncompatibleTransaction(transactions.map(tx => tx.asInstanceOf))} match {
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/forge/ForgeMessageBuilder.scala:225]: 1. Do we need to check it here? Can we rely on mempool correctness?
2. The results interpretation seems strange. So, if `nodeView.state.findIncompatibleTransaction` will find a problem tx, it should return "good" result, but it's processed in the opposite way.
We need to have different method name for `findIncompatibleTransaction`

File: sdk/src/main/scala/com/horizen/storage/SidechainHistoryStorage.scala [hunk 4]
```diff
@@ -88,35 +87,25 @@ class SidechainHistoryStorage(storage: Storage, sidechainTransactionsCompanion:
...
-    storage.get(new ByteArrayWrapper(idToBytes(blockId))).asScala.flatMap { baw =>
-      val tryBlock = new SidechainBlockSerializer(sidechainTransactionsCompanion).parseBytesTry(baw.data)
-      tryBlock match {
-        case Failure(e) => log.warn("SidechainHistoryStorage: Failed to parse block bytes from storage.", e)
-        case _ =>
-      }
-      tryBlock.toOption
-    }
+    val blockIdBytes = new ByteArrayWrapper(idToBytes(blockId))
+    storage.get(blockIdBytes).asScala.flatMap(baw => sidechainBlockSerializer.parseBytesTry(baw.data).toOption)
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/storage/SidechainHistoryStorage.scala]: Unfortunately, we have no logs anymore.

**Review by i-Alex — CHANGES_REQUESTED** [review]

File: sdk/src/main/scala/com/horizen/ClosedBoxesZendooMerkleTree.scala [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+  //current implementation of BigLazyMerkleTree require to use freeLazyMerkleTree() for saving current state at the end of the work
+  Runtime.getRuntime.addShutdownHook(new Thread() {
+    override def run(): Unit = {
+      merkleTree.freeLazyMerkleTree()
+    }
+  })
+
+  //workaround until box id is not Poseidon hash
+  private def bytesToFieldElement(bytes: Array[Byte]): FieldElement = {
+    val fieldBytes = new Array[Byte](96)
```
[the remark this answers is not in the document]
Comment (i-Alex) [reply sdk/src/main/scala/com/horizen/ClosedBoxesZendooMerkleTree.scala:42]: temporary solution.
Need to discuss with Daniele, how position is calculated. Why the solution proposed in previous message will always return position 0?

File: sdk/src/main/scala/com/horizen/SidechainState.scala (hunk 1 above)
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/SidechainState.scala:261]: applicationState shall be checked first

File: sdk/src/main/java/com/horizen/transaction/DefaultTransactionIncompatibilityChecker.java [hunk 6]
```diff
@@ -1,20 +1,33 @@
...
 
 public class DefaultTransactionIncompatibilityChecker
     implements TransactionIncompatibilityChecker
 {
+    //@TODO remove it, static function shall be used
+    static Random rnd = new Random();
+    static String statePath = new File(System.getProperty("java.io.tmpdir") + File.separator + "-" + Math.abs(rnd.nextInt())).getAbsolutePath();
+    static String dbPath = new File(System.getProperty("java.io.tmpdir") + File.separator + "-" + Math.abs(rnd.nextInt())).getAbsolutePath();
+    static String cachePath = new File(System.getProperty("java.io.tmpdir") + File.separator + "-" + Math.abs(rnd.nextInt())).getAbsolutePath();
+    private static final ClosedBoxesZendooMerkleTree tree = new ClosedBoxesZendooMerkleTree(statePath, dbPath, cachePath);
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/java/com/horizen/transaction/DefaultTransactionIncompatibilityChecker.java:20]: not yet :(

File: sdk/src/main/java/com/horizen/transaction/DefaultTransactionIncompatibilityChecker.java [hunk 7]
```diff
@@ -28,6 +41,23 @@
...
     }
 
+    private <T extends BoxTransaction> boolean checkOutputBoxesPositionExclusivity(T newTx, List<T> currentTxs) {
+        List<Long> currentTxsOutputBoxesPosition = currentTxs.stream().flatMap(tx -> {
+            List<Box> boxes = tx.newBoxes();
+            return boxes.stream().map(box -> tree.getPositionForBoxId(new ByteArrayWrapper(box.id())));
+        }).collect(Collectors.toList());
+
+        List<Box> newTxsBoxes = newTx.newBoxes();
+        List<Long> newTxOutputBoxesPosition = newTxsBoxes.stream().map(box -> tree.getPositionForBoxId(new ByteArrayWrapper(box.id()))).collect(Collectors.toList());
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/java/com/horizen/transaction/DefaultTransactionIncompatibilityChecker.java:51]: done

File: sdk/src/main/java/com/horizen/transaction/OutputDataSource.java [hunk 8]
```diff
@@ -0,0 +1,9 @@
+package com.horizen.transaction;
+
+import com.horizen.box.data.NoncedBoxData;
+
+import java.util.List;
+
+public interface OutputDataSource {
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/java/com/horizen/transaction/OutputDataSource.java:7]: nope

File: sdk/src/main/scala/com/horizen/CachedMerkleTree.scala [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+        addCache(position) = fieldElement
+        true
+      case (true, false, true) =>
+        removeCache.remove(position)
+        true
+      case (true, false, false) =>
+        false
+      case (false, true, false) =>
+        false
+      case _ => false
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/CachedMerkleTree.scala:47]: done

File: sdk/src/main/scala/com/horizen/ClosedBoxesZendooMerkleTree.scala [hunk 10]
```diff
@@ -0,0 +1,175 @@
...
+    transaction.boxIdsToOpen().asScala.map(bytesToFieldElement(_)).toSet
+
+  private def getTransactionOutputFieldElements(transaction: SidechainTypes#SCBT): Set[FieldElement] =
+    transaction.newBoxes().asScala.view.map(_.id()).map(bytesToFieldElement).toSet
+
+  private def elementsCouldBeAdded(fieldElements: Set[FieldElement]): Boolean =
+    fieldElements.map(element => merkleTree.getPosition(element)).forall(merkleTree.isPositionEmpty)
+
+  private def elementsCouldBeRemoved(fieldElements: Set[FieldElement]): Boolean =
+    fieldElements.map(element => merkleTree.getPosition(element)).forall(element => !merkleTree.isPositionEmpty(element))
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/ClosedBoxesZendooMerkleTree.scala:61]: renamed to `fieldElement`

File: sdk/src/main/scala/com/horizen/ClosedBoxesZendooMerkleTree.scala [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+  }
+
+  def validateTransactions(transactions: Iterable[SidechainTypes#SCBT]): Option[SidechainTypes#SCBT] = {
+    transactions.foldLeft(new CachedMerkleTree(merkleTree)) {
+      case (cachedMerkleTree: CachedMerkleTree, transaction: SidechainTypes#SCBT) =>
+        val removedFieldElements = getTransactionInputFieldElements(transaction)
+        val addedFieldElements = getTransactionOutputFieldElements(transaction)
+
+        //try to get new cached tree with applied field elements
+        cachedMerkleTree.getUpdatedCachedMerkleTree(removedFieldElements, addedFieldElements) match {
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/ClosedBoxesZendooMerkleTree.scala:78]: renamed to `tryToGetUpdatedCachedMerkleTree`

File: sdk/src/main/scala/com/horizen/ClosedBoxesZendooMerkleTree.scala [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+  private def updateMerkleTree(toRemoveElements: Iterable[FieldElement], toAddElements: Iterable[FieldElement]): Unit = {
+    val toRemovePositions = toRemoveElements.map(merkleTree.getPosition).toArray
+    val toAdd = toAddElements.toList.asJava
+    val toAddPositions = toAddElements.map(element => merkleTree.getPosition(element))
+
+    log.info(s"update merkle tree positions: ${toRemovePositions.mkString("Remove(", ", ", ")")}, add ${toAddPositions.mkString("Add(", ", ", ")")}")
+
+    if (toRemovePositions.nonEmpty) {
+      require(toRemovePositions.forall(!merkleTree.isPositionEmpty(_)), "Position is empty in UTXO merkle tree")
+      //box to add could be placed at the same leaf which are deleted in the same transaction
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/ClosedBoxesZendooMerkleTree.scala:120]: moved before if's due it explain some requirement

File: sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala [hunk 13]
```diff
@@ -189,7 +189,7 @@ class SidechainNodeViewHolder(sidechainSettings: SidechainSettings,
   }
 
   // This method is actually a copy-paste of parent NodeViewHolder.updateState method.
-  // The difference is that State is updated together with Wallet.
+  // The difference is that State is updated together with Wallet, also State receive list of blocks to be removed
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/SidechainNodeViewHolder.scala:192]: done

File: sdk/src/main/scala/com/horizen/SidechainState.scala [hunk 14]
```diff
@@ -106,22 +110,44 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage, val
...
 
   private def validateBlockTransactionsMutuality(mod: SidechainBlock): Unit = {
+    validateTransactionIdExclusivity(mod)
+
+    validateSpentBoxesExclusivity(mod)
+
+    validateNewBoxPositionExclusivity(mod)
+  }
+
+  private def validateTransactionIdExclusivity(mod: SidechainBlock): Unit = {
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/SidechainState.scala:128]: done

File: sdk/src/main/scala/com/horizen/SidechainState.scala [hunk 15]
```diff
@@ -106,22 +110,44 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage, val
...
+  }
 
+  private def validateSpentBoxesExclusivity(mod: SidechainBlock): Unit = {
     val allInputBoxesIds: Seq[ByteArrayWrapper] = mod.transactions.flatMap(tx => tx.boxIdsToOpen().asScala)
     if (allInputBoxesIds.size != allInputBoxesIds.toSet.size) {
       throw new IllegalArgumentException(s"Block ${mod.id} contains duplicated input boxes to open")
     }
   }
 
+  private def validateNewBoxPositionExclusivity(mod: SidechainBlock): Unit = {
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/SidechainState.scala:142]: done

[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/SidechainState.scala:142]: yes, but this is some kind of fast check to not involve closed box merkle tree for checking which could take some time.

File: sdk/src/main/scala/com/horizen/SidechainState.scala [hunk 16]
```diff
@@ -174,6 +200,11 @@ class SidechainState private[horizen] (stateStorage: SidechainStateStorage, val
     }
   }
 
+  def findIncompatibleTransaction(transactions: Iterable[SidechainTypes#SCBT]): Option[SidechainTypes#SCBT] = {
```
[the remark this answers is not in the document]
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/SidechainState.scala:203]: need that function for a forger, for now, also I would like to keep in `SidechainState.validate` just functions calls even maybe ` mod.transactions.foreach(tx => validate(tx).get)` shall be moved to function

File: sdk/src/main/scala/com/horizen/SidechainState.scala (hunk 1 above)
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/SidechainState.scala:261]: done

File: sdk/src/main/scala/com/horizen/SidechainState.scala (hunk 2 above)
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/SidechainState.scala:322]: done.

File: sdk/src/main/scala/com/horizen/forge/ForgeMessageBuilder.scala (hunk 3 above)
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/forge/ForgeMessageBuilder.scala:225]: done

File: sdk/src/main/java/com/horizen/transaction/DefaultTransactionIncompatibilityChecker.java [hunk 17]
```diff
@@ -42,20 +42,23 @@
...
-        List<Long> currentTxsOutputBoxesPosition = currentTxs.stream().flatMap(tx -> {
+        List<Long> currentTxsOutputBoxesPositionList = currentTxs.stream().flatMap(tx -> {
             List<Box> boxes = tx.newBoxes();
             return boxes.stream().map(box -> tree.getPositionForBoxId(new ByteArrayWrapper(box.id())));
         }).collect(Collectors.toList());
 
-        List<Box> newTxsBoxes = newTx.newBoxes();
-        List<Long> newTxOutputBoxesPosition = newTxsBoxes.stream().map(box -> tree.getPositionForBoxId(new ByteArrayWrapper(box.id()))).collect(Collectors.toList());
+        Set<Long> currentTxsOutputBoxesPositionSet = new HashSet<>(currentTxsOutputBoxesPositionList);
+        if (currentTxsOutputBoxesPositionList.size() != currentTxsOutputBoxesPositionSet.size()) {
```
Comment (i-Alex) [inline sdk/src/main/java/com/horizen/transaction/DefaultTransactionIncompatibilityChecker.java:51]: this case means, that mempool was inconsistent even before. Should never happen. Let's assert here.

File: sdk/src/main/scala/com/horizen/ClosedBoxesZendooMerkleTree.scala [hunk 18]
```diff
@@ -55,10 +55,10 @@ class ClosedBoxesZendooMerkleTree(val statePath: String, val dbPath: String, val
     transaction.newBoxes().asScala.view.map(_.id()).map(bytesToFieldElement).toSet
 
   private def elementsCouldBeAdded(fieldElements: Set[FieldElement]): Boolean =
-    fieldElements.map(element => merkleTree.getPosition(element)).forall(merkleTree.isPositionEmpty)
+    fieldElements.map(fieldElement => merkleTree.getPosition(fieldElement)).forall(merkleTree.isPositionEmpty)
 
   private def elementsCouldBeRemoved(fieldElements: Set[FieldElement]): Boolean =
-    fieldElements.map(element => merkleTree.getPosition(element)).forall(element => !merkleTree.isPositionEmpty(element))
+    fieldElements.map(fieldElement => merkleTree.getPosition(fieldElement)).forall(element => !merkleTree.isPositionEmpty(element))
```
Comment (i-Alex) [inline sdk/src/main/scala/com/horizen/ClosedBoxesZendooMerkleTree.scala:61]: in previous comment, I meant the name of `element` in `.forall(element => !merkleTree.isPositionEmpty(element))`.
we should rename it to `position`

**Review by i-Alex — CHANGES_REQUESTED** [review]
The code is fine, just a few minor changes. I still need to review the tests.

File: sdk/src/main/scala/com/horizen/ClosedBoxesZendooMerkleTree.scala (hunk 18 above)
Comment (oaleksandrov-ip) [reply sdk/src/main/scala/com/horizen/ClosedBoxesZendooMerkleTree.scala:61]: done

File: sdk/src/main/java/com/horizen/transaction/DefaultTransactionIncompatibilityChecker.java (hunk 17 above)
Comment (oaleksandrov-ip) [reply sdk/src/main/java/com/horizen/transaction/DefaultTransactionIncompatibilityChecker.java:51]: done

**Pull request closed without merging**